blog/content/post/2024-03-15-uefi-sb-implementation-mpl.md

```diff
@@ -0,0 +1,385 @@
+Yocto layer.
+
+## Integrating Secure Boot into a Yocto layer
+
+Integrating UEFI Secure Boot into an existing Yocto layer is possible by using
+the [meta-secure-core](https://github.com/Wind-River/meta-secure-core) layer in
+your build. Its sublayer – `meta-efi-secure-boot` introduces mechanisms that
+allow verifying various files used in the boot process. It offers two secure
+boot technologies:
+
+- UEFI Secure Boot, which verifies images loaded by UEFI firmware against
+certificates
+- MOK (Machine Owner Key) Secure Boot, which extends UEFI SB by
+adding the shim bootloader.
+
+### MOK Secure Boot
+
+The MOK SB boot flow looks like the following:
+
+```.
+UEFI firmware boot manager (UEFI Secure Boot enabled) ->
+    shim (verified by a DB certificate) ->
+        SELoader (ditto) ->
+            grub (ditto) ->
+                grub.cfg (ditto)
+                kernel (ditto)
+                initramfs (ditto)
+```
+
+Each bootloader will check the integrity of the next-level bootloader before
+launching it.
+
+UEFI firmware will validate the integrity of shim with a certificate in
+DB and then load it.
+
+Shim is the first-stage bootloader. It is trusted by the UEFI Firmware, which
+allows it to be booted on any system with enabled SB. Shim introduces the
+concept of Machine Owner Keys (MOKs), which are user-added keys that can be used
+to sign bootloaders or kernels. Shim will validate and load the SELoader, which
+has to be signed by a private key corresponding to a DB certificate, the shim
+certificate, the vendor certificate, or a MOK certificate.
+
+The SELoader is a second-stage bootloader, which authenticates grub
+configuration files, the kernel, and initramfs.
+
+The layer also introduces GRUB Lockdown, which prevents modifying the kernel
+command line or loading unsigned boot components. When Secure Boot is enabled,
+the user can only access the command line if the user authentication is enabled,
+in which case it is protected by a password.
+
+The combination of the mechanisms vastly increases the level of security and
+trustworthiness of the boot process, however in this section we will focus on
+implementing regular UEFI SB.
+
+### Implementation
+
+To integrate the mechanisms available in `meta-secure-core` you need to
+integrate that layer into your build. We use
+[kas-container](https://github.com/siemens/kas/blob/master/kas-container) to set
+up bitbake projects.
+
+In your kas configuration file add the following to the `repos` section:
+
+```yaml
+meta-secure-core:
+url: https://github.com/Wind-River/meta-secure-core.git
+refspec: 8dc9f1b4a735eccee65a8896760e473e110d147e
+layers:
+    meta-efi-secure-boot:
+    meta:
+    meta-signing-key:
+```
+
+Define `meta-secure-core` variables in your layer's `local.conf`:
+
+```conf
+# UEFI Secure Boot variables
+# Use only UEFI Secure Boot without MOK Secure Boot
+UEFI_SELOADER = "0"
+GRUB_SIGN_VERIFY = "1"
+UEFI_SB = "1"
```

Comment:
@PLangowski also

```
MOK_SB = "0"
```

is needed to use UEFI Secure Boot without the shim.

Reply:
Added 71a27177c0752fe29b2e38ce670b194e9ef094b3